doc/developer/platform/ux.md

```diff
@@ -299,17 +299,63 @@ The HTTPS server exposes two APIs:
     additional control over query execution and result fetching, a la the
     [Snowflake SQL REST API].
 
-    TODO: fully describe the output format.
+    ### Output format
 
-    TODO: design support for binding parameters.
+    The output format is a JSON object with one key, `results`, whose values is
+    an array of serialized `adapter::command::SimpleResult`, representing all of
+    the statements' executions.
+
+    The serialized `adapter::command::SimpleResult` values are representing in JSON as:
+
+    `SimpleResult` value | JSON value
+    ---------------------|------------
+    `Rows` | `{"rows": <2D array of JSON-ified results>, "col_names": <array of text>}`
+    `Error` | `{"error": <Error string from execution>}`
+    `Ok` | `{ "ok": <tag>}`
+
+    ### Binding parameters
+
+    You can include parameters alongside the query by including on the query a
+    key `"parameters"` whose value is an array of objects of the form:
+
+    ```json
+    {
+      "type": "<resolvable type name>",
+      "value": "<JSON-ified value>"
+    }
+    ```
+
+    You can then reference these parameters in the query as you would with PG,
+    i.e. using 1-indexed references to the parameters array, prefaced by a `$`. e.g.
+
+    ```json
+    {
+      "sql": "INSERT INTO foo VALUES ($1);",
+      "parameters": [{"type": "int", "value": 1}]
+    }
+    ```
 
   * `ws://<ENDPOINT>/query?query=QUERY` streams the result of a single query
     over the WebSockets protocol.
 
     This API is particularly useful for listening to the results of `TAIL`
     queries from a browser.
 
-    TODO: fully describe the output format.
+    ### Output
+
+    A Websocket request outputs the following:
```

Comment:
If I were trying to write a thing for consuming websocket results, I'd want some additional information regarding guarantees, order of messages, and termination. Of the four types returned here, do they always appear in some specific order or combination? For example: 

- the first thing will always be either `ok` or `error`
- if it's `error` then the stream terminates
- if it's `ok` and there are no rows returned, the stream terminates
- if it's `ok` and there are rows, a `col_names` is returned, followed by any number of `rows` results (including 0), after which the stream terminates.

In `ok` case, we need a way to explicitly differentiate between "i'm all done and there are no more rows" and "there has been a connection error from the server so there might have been more rows, dunno". I don't know enough about the details of ws right now to know if ws exposes those differences or not. If not, then we should include an explicit end message so rows consumers have a certainty that they got all their results.

Another idea: the `ok` message could include an optional `col_names` field that, if present, indicates rows can come.

Reply:
Clarified ordering is:
1. `col_names` if values will get returned, then any `rows` returned, of which there may be many.
2. `ok` or `error` indicating end of processing; `col_names` never precedes `error`.